README.md
# OmniAuth Latvija.lv

> **Autentifikācijas serviss.** Ar autentifikācijas servisa palīdzību iestādes var ērti savos pakalpojumu portālos vai reģistros autentificēt pakalpojuma lietotāju atbilstoši visiem drošības standartiem. Iestādei vairs nav nepieciešams veidot jaunus autentifikācijas mehānismus un juridiski slēgt vienošanos ar katru no autentifikācijas iespējas nodrošinātājiem. Šobrīd pieejamos autentifikācijas līdzekļus var apskatīties portālā [www.latvija.lv](http://www.latvija.lv/) sadaļā „E-pakalpojumi”.
>
> -- [VRAA E-pakalpojumi](http://www.vraa.gov.lv/lv/epakalpojumi/viss/)

Provides the following authentication types:

* E-paraksts
* Swedbank
* SEB banka
* Online bank
* Citadele
* Norvik banka
* PrivatBank
* eID
* Lattelecom Mobile ID

## Installation

```ruby
gem 'omniauth-latvija'
```

## Usage

`OmniAuth::Strategies::Latvija` is simply a Rack middleware. Read the OmniAuth 1.x docs for detailed instructions: https://github.com/intridea/omniauth.

Here's a quick example, adding the middleware to a Rails app in `config/initializers/omniauth.rb`:

```ruby
Rails.application.config.middleware.use OmniAuth::Builder do
  provider :latvija, {
    endpoint:    "https://epaktv.vraa.gov.lv/IVIS.LVP.STS/Default.aspx",
    certificate: File.read("/path/to/cert.pem"),
    private_key: File.read("/path/to/private_key.pem"), # mandatory, if the response is encrypted
    realm:       "urn:federation:example.com"
  }
end
```


## Auth Hash

Here's an example hash available in `request.env['omniauth.auth']`

```ruby
{
  provider: 'latvija',
  uid: 'PK:12345612345',
  info: {
    name: 'JANIS BERZINS',
    first_name: 'JANIS',
    last_name: 'BERZINS',
    private_personal_identifier: '12345612345'
  },
  extra: {
    raw_info: {
      givenname: 'JANIS',
      surname: 'BERZINS',
      privatepersonalidentifier: '12345612345',
      historical_privatepersonalidentifier: [],
      not_valid_before: '2019-05-09T07:29:41Z',
      not_valid_on_or_after: '2019-05-09T08:29:41Z'
    },
    authentication_method: 'URN:IVIS:100001:AM.BANK-SWED',
    original_issuer: 'Swedbanka',
    legacy_uids: ['JANIS BERZINS, 12345612345']
  }
}
```

## References

* http://docs.oasis-open.org/wsfed/federation/v1.2/os/ws-federation-1.2-spec-os.html
* http://msdn.microsoft.com/en-us/library/bb498017.aspx
* http://msdn.microsoft.com/en-us/library/bb608217.aspx
* https://github.com/onelogin/ruby-saml


CHANGELOG.md
## 6.2.0

- Add original issuer to response #11

## 6.1.0

- Use single private identifier as primary #10

## 6.0.0

- Add SHA256 support #9

## 5.0.0

- **BREAKING CHANGE**: use the identifier returned by latvija.lv as auth UID, to prevent surname changes resetting user identity. Without changes in caller code, existing users will not be able to log in. Previously used versions of UID can be found under `extra.legacy_uids` key in auth response.
- Return correct personal identifier in cases when person has had it changed. Historical identifiers, if any, can be found under `extra.raw_info.historical_privatepersonalidentifier` key in auth response.

## 4.0.0

- Check authentication expirity #6

## 3.0.0

- Normalize omniauth attribute names #5

## 2.0.0

- Support for encrypted responses #4
- Switch XML parsing and canonicalization to Nokogiri #4
- Switch to ruby 2.1 Syntax (removed support for Ruby <2.0) #4

## 1.1.1

- Added 'wfresh' parameter support in request phase #2

## 1.1.0

- Pass back authentication_method #1

## 1.0.0

- Initial version


spec/spec_helper.rb
$:.unshift File.dirname(__FILE__) + '/../lib'

require 'rubygems'
require 'bundler/setup'

require 'simplecov'
SimpleCov.start

require 'rspec'
require 'rack/test'
require 'timecop'

require 'omniauth'
require 'omniauth/test'
require 'omniauth-latvija'

RSpec.configure do |config|
  config.include Rack::Test::Methods
  config.extend  OmniAuth::Test::StrategyMacros, :type => :strategy
end


spec/omniauth/strategies/latvija_spec.rb
require File.expand_path('../../../spec_helper', __FILE__)

describe OmniAuth::Strategies::Latvija, :type => :strategy do
  include OmniAuth::Test::StrategyTestCase

  let(:fixtures_valid_from_inclusive) { Time.parse '2017-10-17T14:56:04.831Z' }
  let(:fixtures_valid_to_not_inclusive) { Time.parse '2017-10-17T18:56:04.831Z' }
  let(:freeze_time_at) { fixtures_valid_from_inclusive }

  let(:certificate) { File.read('spec/fixtures/private/cert.pem') }
  let(:private_key) { File.read('spec/fixtures/private/key.pem') }
  let(:wresult_encrypted) { File.read('spec/fixtures/private/wresult_encrypted.xml') }
  let(:wresult_decrypted) { File.read('spec/fixtures/private/wresult_decrypted.xml') }

  def strategy
    [ OmniAuth::Strategies::Latvija,
      { certificate: certificate,
        endpoint: "https://epaktv.vraa.gov.lv/IVIS.LVP.STS/Default.aspx",
        realm: "urn:federation:example.com",
        private_key: private_key
      }
    ]
  end

  around do |example|
    Timecop.freeze(freeze_time_at) do
      example.run
    end
  end

  describe '/auth/latvija' do
    it 'should redirect to latvija.lv authentication screen' do
      get '/auth/latvija'
      last_response.should be_redirect
      # puts last_response.headers['Location'].inspect
      last_response.headers['Location'].should match %r{https://epaktv.vraa.gov.lv/IVIS.LVP.STS/Default.aspx\?wa=wsignin1.0&wct=([^&]+)&wtrealm=urn%3Afederation%3Aexample.com&wreply=http%3A%2F%2Fexample.org%2Fauth%2Flatvija%2Fcallback&wctx=http%3A%2F%2Fexample.org%2Fauth%2Flatvija%2Fcallback&wreq=%3Ctrust}
    end

    it 'should gather user data after successful authentication using Swedbank' do
      post '/auth/latvija/callback', {
        :wa => "wsignin1.0",
        :wctx => "http://example.org/auth/latvija/callback",
        :wresult => wresult_encrypted
      }

      # puts "omniauth.auth: #{last_request.env['omniauth.auth'].inspect}"
      # puts "omniauth.error: #{last_request.env['omniauth.error'].inspect}"
      # puts "omniauth.error.type: #{last_request.env['omniauth.error.type'].inspect}"
      # puts "omniauth.strategy: #{last_request.env['omniauth.strategy'].inspect}"

      last_request.env['omniauth.error'].should be_nil
      last_request.env['omniauth.auth']['uid'].should be_present
      last_request.env['omniauth.auth']['extra']['raw_info']['givenname'].should be_present
      last_request.env['omniauth.auth']['extra']['raw_info']['surname'].should be_present
      last_request.env['omniauth.auth']['info']['first_name'].should be_present
      last_request.env['omniauth.auth']['info']['last_name'].should be_present
      last_request.env['omniauth.auth']['info']['private_personal_identifier'].should be_present
    end

    it 'should fail after unsuccessful authentication with fingerprint mismatch' do
      invalid_certificate = <<-EOS
        -----BEGIN CERTIFICATE-----
        MIICNDCCAaECEAKtZn5ORf5eV288mBle3cAwDQYJKoZIhvcNAQECBQAwXzELMAkG
        A1UEBhMCVVMxIDAeBgNVBAoTF1JTQSBEYXRhIFNlY3VyaXR5LCBJbmMuMS4wLAYD
        VQQLEyVTZWN1cmUgU2VydmVyIENlcnRpZmljYXRpb24gQXV0aG9yaXR5MB4XDTk0
        MTEwOTAwMDAwMFoXDTEwMDEwNzIzNTk1OVowXzELMAkGA1UEBhMCVVMxIDAeBgNV
        BAoTF1JTQSBEYXRhIFNlY3VyaXR5LCBJbmMuMS4wLAYDVQQLEyVTZWN1cmUgU2Vy
        dmVyIENlcnRpZmljYXRpb24gQXV0aG9yaXR5MIGbMA0GCSqGSIb3DQEBAQUAA4GJ
        ADCBhQJ+AJLOesGugz5aqomDV6wlAXYMra6OLDfO6zV4ZFQD5YRAUcm/jwjiioII
        0haGN1XpsSECrXZogZoFokvJSyVmIlZsiAeP94FZbYQHZXATcXY+m3dM41CJVphI
        uR2nKRoTLkoRWZweFdVJVCxzOmmCsZc5nG1wZ0jl3S3WyB57AgMBAAEwDQYJKoZI
        hvcNAQECBQADfgBl3X7hsuyw4jrg7HFGmhkRuNPHoLQDQCYCPgmc4RKz0Vr2N6W3
        YQO2WxZpO8ZECAyIUwxrl0nHPjXcbLm7qt9cuzovk2C2qUtN8iD3zV9/ZHuO3ABc
        1/p3yjkWWW8O6tO1g39NTUJWdrTJXwT4OPjr0l91X817/OWOgHz8UA==
        -----END CERTIFICATE-----
      EOS

      post '/auth/latvija/callback', {
        :wa => "wsignin1.0",
        :wctx => "http://example.org/auth/latvija/callback",
        :wresult => wresult_decrypted.sub(
          certificate.sub("-----BEGIN CERTIFICATE-----", "").sub("-----END CERTIFICATE-----", "").gsub("\n", ""),
          invalid_certificate.sub("-----BEGIN CERTIFICATE-----", "").sub("-----END CERTIFICATE-----", "").gsub("\n", "")
        )
      }

      last_request.env['omniauth.error'].message.should == "Fingerprint mismatch"
    end

    it 'should fail after unsuccessful authentication when signature digest is invalid' do
      post '/auth/latvija/callback', {
        :wa => "wsignin1.0",
        :wctx => "http://example.org/auth/latvija/callback",
        :wresult => wresult_decrypted.sub(
          /<DigestValue>[^<]+<\/DigestValue>/,
          '<DigestValue>0FA</DigestValue>'
        )
      }

      last_request.env['omniauth.error'].message.should == "Digest mismatch"
      last_request.env['omniauth.auth'].should be_nil
    end

    it 'should fail after unsuccessful authentication when signature key is invalid' do
      post '/auth/latvija/callback', {
        :wa => "wsignin1.0",
        :wctx => "http://example.org/auth/latvija/callback",
        :wresult => wresult_decrypted.sub(
          /<SignatureValue>[^<]+<\/SignatureValue>/,
          '<SignatureValue>0FA</SignatureValue>'
        )
      }

      last_request.env['omniauth.error'].message.should == "Key validation error"
      last_request.env['omniauth.auth'].should be_nil
    end

    it 'should fail after unsuccessful authentication when wresult is empty' do
      post '/auth/latvija/callback', {
        :wa => "wsignin1.0",
        :wctx => "http://example.org/auth/latvija/callback"
      }

      last_request.env['omniauth.error.type'].should == :invalid_response
      last_request.env['omniauth.auth'].should be_nil
    end

    context 'timestamp validation' do
      context 'when response is still valid' do
        let(:freeze_time_at) { fixtures_valid_to_not_inclusive - 1.second }

        it 'should not fail' do
          post '/auth/latvija/callback', {
            :wa => "wsignin1.0",
            :wctx => "http://example.org/auth/latvija/callback",
            :wresult => wresult_decrypted
          }

          last_request.env['omniauth.error'].should be_nil
          last_request.env['omniauth.auth'].should be_present
          last_request.env['omniauth.auth']['extra']['raw_info']['not_valid_before'].should be_present
          last_request.env['omniauth.auth']['extra']['raw_info']['not_valid_on_or_after'].should be_present
        end
      end

      context 'when response is no longer valid' do
        let(:freeze_time_at) { fixtures_valid_to_not_inclusive }

        it 'should fail' do
          post '/auth/latvija/callback', {
            :wa => "wsignin1.0",
            :wctx => "http://example.org/auth/latvija/callback",
            :wresult => wresult_decrypted
          }

          last_request.env['omniauth.error'].message.should == 'Current time is on or after NotOnOrAfter condition'
          last_request.env['omniauth.auth'].should be_nil
        end
      end
    end

    context 'when response has multiple private identifiers' do
      let(:wresult_decrypted) { File.read('spec/fixtures/wresult_multi_personal_codes_decrypted.xml') }

      before(:each) do
        allow_any_instance_of(OmniAuth::Strategies::Latvija::SignedDocument).to receive(:validate!).and_return(true)
      end

      let(:response) do
        post '/auth/latvija/callback', {
          :wa => "wsignin1.0",
          :wctx => "http://example.org/auth/latvija/callback",
          :wresult => wresult_decrypted
        }

        last_request.env['omniauth.auth']
      end

      it 'should return first name' do
        expect(response.dig('info', 'first_name')).to eq('ODS')
      end

      it 'should return last name' do
        expect(response.dig('info', 'last_name')).to eq('KNISLIS')
      end

      it 'should return combined name' do
        expect(response.dig('info', 'name')).to eq('ODS KNISLIS')
      end

      it 'should return primary personal code, as specified by OriginalIssuer param' do
        expect(response.dig('info', 'private_personal_identifier')).to eq('32345678901')
      end

      it 'should return any historical personal codes in extra info' do
        expect(response.dig('extra', 'raw_info', 'historical_privatepersonalidentifier')).to match_array(['12345678901'])
      end

      it 'should return NameIdentifier property as the auth UID' do
        expect(response.dig('uid')).to eq('PK:32345678901')
      end

      it 'should return legacy UIDs for gem version <= 4.0' do
        expect(response.dig('extra', 'legacy_uids')).to include('ODS KNISLIS, 32345678901')
        expect(response.dig('extra', 'legacy_uids')).to include('ODS KNISLIS, 12345678901')
      end

      it 'should return legacy UIDs for persons that have changed their identifiers' do
        expect(response.dig('extra', 'legacy_uids')).to include('PK:12345678901')
      end
    end

    context 'when response has single private identifier without original issuer parameter' do
      let(:wresult_decrypted) { File.read('spec/fixtures/wresult_single_personal_code_without_issuer_decrypted.xml') }

      before(:each) do
        allow_any_instance_of(OmniAuth::Strategies::Latvija::SignedDocument).to receive(:validate!).and_return(true)
      end

      let(:response) do
        post '/auth/latvija/callback', {
          :wa => "wsignin1.0",
          :wctx => "http://example.org/auth/latvija/callback",
          :wresult => wresult_decrypted
        }

        last_request.env['omniauth.auth']
      end

      it 'should return primary personal code' do
        expect(response.dig('info', 'private_personal_identifier')).to eq('32345678901')
      end

      it 'should not return historical personal codes in extra info' do
        expect(response.dig('extra', 'raw_info', 'historical_privatepersonalidentifier')).to be_empty
      end

      it 'should return NameIdentifier property as the auth UID' do
        expect(response.dig('uid')).to eq('PK:32345678901')
      end
    end
  end
end


lib/omniauth/strategies/latvija/response.rb
module OmniAuth::Strategies
  class Latvija
    class Response
      ASSERTION = 'urn:oasis:names:tc:SAML:1.0:assertion'.freeze

      attr_accessor :options, :response

      def initialize(response, **options)
        raise ArgumentError, 'Response cannot be nil' if response.nil?
        @options  = options
        @response = response
        @document = OmniAuth::Strategies::Latvija::SignedDocument.new(response, private_key: options[:private_key])
      end

      def validate!
        @document.validate!(fingerprint) && validate_conditions!
      end

      def xml
        @document.nokogiri_xml
      end

      def authentication_method
        @authentication_method ||= begin
          xml.xpath('//saml:AuthenticationStatement', saml: ASSERTION).attribute('AuthenticationMethod')
        end
      end

      def original_issuer
        @original_issuer ||= begin
          xml.xpath("//saml:Attribute[@AttributeName='givenname']", saml: ASSERTION).attribute('OriginalIssuer')
        end
      end

      def name_identifier
        @name_identifier ||= begin
          xml.xpath('//saml:AuthenticationStatement/saml:Subject/saml:NameIdentifier', saml: ASSERTION).text()
        end
      end

      # A hash of all the attributes with the response.
      # Assuming there is only one value for each key
      def attributes
        @attributes ||= begin
          attrs = {
            'not_valid_before' => not_valid_before,
            'not_valid_on_or_after' => not_valid_on_or_after,
            'historical_privatepersonalidentifier' => []
          }

          stmt_elements = xml.xpath('//saml:Attribute', saml: ASSERTION)

          return attrs if stmt_elements.nil?

          identifiers = stmt_elements.xpath("//saml:Attribute[@AttributeName='privatepersonalidentifier']", saml: ASSERTION)

          stmt_elements.each_with_object(attrs) do |element, result|
            name = element.attribute('AttributeName').value
            value = element.text

            case name
            when 'privatepersonalidentifier' # person can change their identifier, service will return all the versions
              if identifiers.length == 1 || element.attribute('OriginalIssuer') # this is the primary identifier, as returned by third party auth service
                result[name] = value
              else
                result['historical_privatepersonalidentifier'] << value
              end
            else
              result[name] = value
            end
          end
        end
      end

      private

      def fingerprint
        cert = OpenSSL::X509::Certificate.new(options[:certificate])
        Digest::SHA256.hexdigest(cert.to_der).upcase.scan(/../).join(':')
      end

      def conditions_tag
        @conditions_tag ||= xml.xpath('//saml:Conditions', saml: ASSERTION)
      end

      def not_valid_before
        @not_valid_before ||= conditions_tag.attribute('NotBefore').value
      end

      def not_valid_on_or_after
        @not_valid_on_or_after ||= conditions_tag.attribute('NotOnOrAfter').value
      end

      def validate_conditions!
        if not_valid_on_or_after.present? && Time.current < Time.parse(not_valid_on_or_after)
          true
        else
          raise ValidationError, 'Current time is on or after NotOnOrAfter condition'
        end
      end
    end
  end
end


lib/omniauth/strategies/latvija/signed_document.rb
# The contents of this file are subject to the terms
# of the Common Development and Distribution License
# (the License). You may not use this file except in
# compliance with the License.
#
# You can obtain a copy of the License at
# https://opensso.dev.java.net/public/CDDLv1.0.html or
# opensso/legal/CDDLv1.0.txt
# See the License for the specific language governing
# permission and limitations under the License.
#
# When distributing Covered Code, include this CDDL
# Header Notice in each file and include the License file
# at opensso/legal/CDDLv1.0.txt.
# If applicable, add the following below the CDDL Header,
# with the fields enclosed by brackets [] replaced by
# your own identifying information:
# 'Portions Copyrighted [year] [name of copyright owner]'
#
# $Id: xml_sec.rb,v 1.6 2007/10/24 00:28:41 todddd Exp $
#
# Copyright 2007 Sun Microsystems Inc. All Rights Reserved
# Portions Copyrighted 2007 Todd W Saxton.

module OmniAuth::Strategies
  class Latvija
    class SignedDocument
      DSIG = 'http://www.w3.org/2000/09/xmldsig#'.freeze
      XENC = 'http://www.w3.org/2001/04/xmlenc#'.freeze
      CANON_MODE = Nokogiri::XML::XML_C14N_EXCLUSIVE_1_0

      def initialize(response, **opts)
        @response = Nokogiri::XML.parse(response, &:noblanks)
        return unless encrypted?
        decryptor = OmniAuth::Strategies::Latvija::Decryptor.new(response, opts[:private_key])
        decrypted_response = decryptor.decrypt
        @response = Nokogiri::XML.parse(decrypted_response, &:noblanks)
      end

      def validate!(idp_cert_fingerprint)
        validate_fingerprint!(idp_cert_fingerprint)
        sig_element = @response.xpath('//xmlns:Signature', xmlns: DSIG)

        validate_digest!(sig_element)
        validate_signature!(sig_element)
        true
      end

      def nokogiri_xml
        @response
      end

      private

      def encrypted?
        @response.xpath('//xenc:EncryptedData', 'xmlns:xenc' => XENC).any?
      end

      def certificate
        @certificate ||= begin
          base64_cert = @response.xpath('//xmlns:X509Certificate', xmlns: DSIG).text
          cert_text   = Base64.decode64(base64_cert)
          OpenSSL::X509::Certificate.new(cert_text)
        end
      end

      def digest_method_class(reference)
        value = reference.xpath('.//xmlns:DigestMethod', xmlns: DSIG).attribute('Algorithm').value
        value == "#{DSIG}sha1" ? Digest::SHA1 : Digest::SHA256
      end

      def signature_method_class(sig_element)
        value = sig_element.xpath('.//xmlns:SignatureMethod', xmlns: DSIG).attribute('Algorithm').value
        value == "#{DSIG}rsa-sha1" ? OpenSSL::Digest::SHA1 : OpenSSL::Digest::SHA256
      end

      def validate_fingerprint!(idp_cert_fingerprint)
        fingerprint = Digest::SHA256.hexdigest(certificate.to_der)
        if fingerprint != idp_cert_fingerprint.gsub(/[^a-zA-Z0-9]/, '').downcase
          raise ValidationError, 'Fingerprint mismatch'
        end
      end

      def validate_digest!(sig_element)
        response_without_signature = @response.dup
        response_without_signature.xpath('//xmlns:Signature', xmlns: DSIG).remove

        sig_element.xpath('.//xmlns:Reference', xmlns: DSIG).each do |ref|
          uri            = ref.attribute('URI').value
          hashed_element = response_without_signature.
            at_xpath("//*[@AssertionID='#{uri[1, uri.size]}']").
            canonicalize(CANON_MODE)
          hash           = Base64.encode64(digest_method_class(ref).digest(hashed_element)).chomp
          digest_value   = ref.xpath('.//xmlns:DigestValue', xmlns: DSIG).text

          raise ValidationError, 'Digest mismatch' if hash != digest_value
        end
      end

      def validate_signature!(sig_element)
        signed_info_element = sig_element.
          at_xpath('.//xmlns:SignedInfo', xmlns: DSIG).
          canonicalize(CANON_MODE)
        base64_signature    = sig_element.xpath('.//xmlns:SignatureValue', xmlns: DSIG).text
        signature           = Base64.decode64(base64_signature)

        unless certificate.public_key.verify(signature_method_class(sig_element).new, signature, signed_info_element)
          raise ValidationError, 'Key validation error'
        end
      end
    end
  end
end


lib/omniauth/strategies/latvija.rb
require 'time'
require 'openssl'
require 'digest/sha1'
require 'digest/sha2'
require 'xmlenc'
require 'nokogiri'
require 'omniauth/strategies/latvija/response'
require 'omniauth/strategies/latvija/decryptor'
require 'omniauth/strategies/latvija/signed_document'

module OmniAuth::Strategies
  #
  # Authenticate with Latvija.lv.
  #
  # @example Basic Rails Usage
  #
  #  Add this to config/initializers/omniauth.rb
  #
  #    Rails.application.config.middleware.use OmniAuth::Builder do
  #      provider :latvija, {
  #        endpoint:    "https://epaktv.vraa.gov.lv/IVIS.LVP.STS/Default.aspx",
  #        certificate: File.read("/path/to/cert"),
  #        private:     File.read("/path/to/private_key"),
  #        realm:       "urn:federation:example.com"
  #      }
  #    end
  #
  class Latvija
    include OmniAuth::Strategy
    class ValidationError < StandardError; end

    option :realm, nil
    option :wfresh, false
    option :endpoint, nil
    option :certificate, nil
    option :private_key, nil

    info do
      {
        name: full_name,
        first_name: raw_info['givenname'],
        last_name: raw_info['surname'],
        private_personal_identifier: raw_info['privatepersonalidentifier']
      }
    end

    extra do
      {
        raw_info: raw_info,
        authentication_method: @response.authentication_method,
        original_issuer: @response.original_issuer,
        legacy_uids: legacy_uids
      }
    end

    def request_phase
      params = {
        wa: 'wsignin1.0',
        wct: Time.now.utc.strftime('%Y-%m-%dT%H:%M:%SZ'),
        wtrealm: options[:realm],
        wreply: callback_url,
        wctx: callback_url,
        wreq: '<trust:RequestSecurityToken xmlns:trust="http://docs.oasis-open.org/ws-sx/ws-trust/200512"><trust:Claims xmlns:i="http://schemas.xmlsoap.org/ws/2005/05/identity" Dialect="http://schemas.xmlsoap.org/ws/2005/05/identity"><i:ClaimType Uri="http://docs.oasis-open.org/wsfed/authorization/200706/claims/action" Optional="false" /></trust:Claims><trust:RequestType>http://docs.oasis-open.org/ws-sx/ws-trust/200512/Issue</trust:RequestType></trust:RequestSecurityToken>'
      }
      params[:wfresh] = options[:wfresh] if options[:wfresh]
      query_string = params.collect { |key, value| "#{key}=#{Rack::Utils.escape(value)}" }.join('&')
      redirect "#{options[:endpoint]}?#{query_string}"
    end

    def callback_phase
      if request.params['wresult']
        @response = OmniAuth::Strategies::Latvija::Response.new(
          request.params['wresult'],
          certificate: options[:certificate],
          private_key: options[:private_key]
        )
        @response.validate!
        super
      else
        fail!(:invalid_response)
      end
    rescue Exception => e
      fail!(:invalid_response, e)
    end

    def raw_info
      @response.attributes
    end

    def uid
      @response.name_identifier
    end

    def full_name
      @full_name ||= "#{raw_info['givenname']} #{raw_info['surname']}"
    end

    def legacy_uids
      # UIDs that could have been assigned to this identity by previous versions of the gem, or due to peronal identifier change

      legacy_uids = [
        "#{full_name}, #{raw_info["privatepersonalidentifier"]}" # generated by gem version <= 4.0
      ]

      raw_info.fetch('historical_privatepersonalidentifier', []).each do |historical_identifier|
        legacy_uids << "#{full_name}, #{historical_identifier}" # generated by gem version <= 4.0
        legacy_uids << "PK:#{historical_identifier}" # due to personal identifier change
      end

      legacy_uids
    end
  end
end
